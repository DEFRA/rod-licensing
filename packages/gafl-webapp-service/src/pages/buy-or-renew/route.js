import { BUY_OR_RENEW } from '../../uri.js'
import Joi from 'joi'
import pageRoute from '../../routes/page-route.js'
import { nextPage } from '../../routes/next-page.js'
export const buyOrRenew = {
  buy: 'buy',
  renew: 'renew'
}

export const validator = Joi.object({
  'buy-or-renew': Joi.string()
    .valid('buy-licence', 'renew-licence')
    .required()
}).options({ abortEarly: false, allowUnknown: true })

const route = pageRoute(BUY_OR_RENEW.page, BUY_OR_RENEW.uri, validator, nextPage)

export default route
