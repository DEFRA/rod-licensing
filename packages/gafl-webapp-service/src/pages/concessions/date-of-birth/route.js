import { DATE_OF_BIRTH } from '../../../uri.js'
import Joi from 'joi'
import pageRoute from '../../../routes/page-route.js'
import { validation } from '@defra-fish/business-rules-lib'
import { nextPage } from '../../../routes/next-page.js'

export const validator = payload => {
  const dateOfBirth = `${payload['date-of-birth-year']}-${payload['date-of-birth-month']}-${payload['date-of-birth-day']}`
  Joi.assert(
    { 'date-of-birth': dateOfBirth },
    Joi.object({
      'date-of-birth': validation.contact.createBirthDateValidator(Joi)
    })
  )
}

export const getData = async request => {
  const { isLicenceForYou } = await request.cache().helpers.status.getCurrentPermission()

  return { isLicenceForYou }
}

export default pageRoute(DATE_OF_BIRTH.page, DATE_OF_BIRTH.uri, validator, nextPage, getData)
