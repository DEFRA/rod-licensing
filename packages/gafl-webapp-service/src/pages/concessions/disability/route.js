import { DISABILITY_CONCESSION } from '../../../uri.js'
import pageRoute from '../../../routes/page-route.js'
import Joi from 'joi'
import { validation } from '@defra-fish/business-rules-lib'
import * as concessionHelper from '../../../processors/concession-helper.js'
import { nextPage } from '../../../routes/next-page.js'
import { disabilityConcessionTypes } from './update-transaction.js'

const validator = Joi.object({
  'disability-concession': Joi.string()
    .valid(...Object.values(disabilityConcessionTypes))
    .required(),
  'ni-number': Joi.alternatives().conditional('disability-concession', {
    is: disabilityConcessionTypes.pipDla,
    then: validation.contact.createNationalInsuranceNumberValidator(Joi).required(),
    otherwise: Joi.string().empty('')
  }),
  'blue-badge-number': Joi.alternatives().conditional('disability-concession', {
    is: disabilityConcessionTypes.blueBadge,
    then: Joi.string()
      .max(25)
      .required(),
    otherwise: Joi.string().empty('')
  })
}).options({ abortEarly: false, allowUnknown: true })

export const getData = async request => {
  const permission = await request.cache().helpers.transaction.getCurrentPermission()

  return {
    hasJunior: concessionHelper.hasJunior(permission),
    hasSenior: concessionHelper.hasSenior(permission),
    isLicenceForYou: permission.isLicenceForYou,
    ...disabilityConcessionTypes
  }
}

export default pageRoute(DISABILITY_CONCESSION.page, DISABILITY_CONCESSION.uri, validator, nextPage, getData)
