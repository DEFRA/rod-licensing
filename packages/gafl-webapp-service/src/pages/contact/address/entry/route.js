import { ADDRESS_ENTRY, ADDRESS_LOOKUP } from '../../../../uri.js'
import pageRoute from '../../../../routes/page-route.js'
import Joi from 'joi'
import { validation } from '@defra-fish/business-rules-lib'
import { countries } from '../../../../processors/refdata-helper.js'
import { nextPage } from '../../../../routes/next-page.js'
import { getPronoun } from '../../../../processors/licence-type-display.js'

export const validator = Joi.object({
  premises: validation.contact.createPremisesValidator(Joi),
  street: validation.contact.createStreetValidator(Joi),
  locality: validation.contact.createLocalityValidator(Joi),
  town: validation.contact.createTownValidator(Joi),
  postcode: Joi.alternatives().conditional('country-code', {
    is: Joi.string().pattern(/^GB/),
    then: validation.contact.createUKPostcodeValidator(Joi),
    otherwise: validation.contact.createOverseasPostcodeValidator(Joi)
  }),
  'country-code': Joi.string().required()
}).options({ abortEarly: false, allowUnknown: true })

export const getCountryDropDownOptions = async () => {
  const options = await countries.getAll()
  return options.filter(country => country.code !== 'GB')
}

export const getData = async request => {
  const { addresses, searchTerms } = await request.cache().helpers.addressLookup.getCurrentPermission()
  const { isLicenceForYou } = await request.cache().helpers.transaction.getCurrentPermission()

  const pronoun = getPronoun(isLicenceForYou).possessive

  return {
    pronoun,
    searchTerms: !addresses?.length && searchTerms ? searchTerms : null,
    countries: await getCountryDropDownOptions(),
    lookupPage: ADDRESS_LOOKUP.uri
  }
}

export default pageRoute(ADDRESS_ENTRY.page, ADDRESS_ENTRY.uri, validator, nextPage, getData)
