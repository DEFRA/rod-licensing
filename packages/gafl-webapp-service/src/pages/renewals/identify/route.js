import { IDENTIFY, AUTHENTICATE, NEW_TRANSACTION } from '../../../uri.js'
import pageRoute from '../../../routes/page-route.js'
import Joi from 'joi'
import { validation } from '@defra-fish/business-rules-lib'

import GetDataRedirect from '../../../handlers/get-data-redirect.js'

const getData = async request => {
  // If we are supplied a permission number, validate it or throw 400
  const permission = await request.cache().helpers.status.getCurrentPermission()

  if (permission.referenceNumber) {
    const validatePermissionNumber = validation.permission
      .permissionNumberUniqueComponentValidator(Joi)
      .validate(permission.referenceNumber)
    if (validatePermissionNumber.error) {
      await request.cache().helpers.status.setCurrentPermission({ referenceNumber: null })
      throw new GetDataRedirect(IDENTIFY.uri)
    }
  }

  return {
    uri: { 
      new: NEW_TRANSACTION.uri
    },
    referenceNumber: permission.referenceNumber
  }
}

const schema = Joi.object({
  referenceNumber: validation.permission.permissionNumberUniqueComponentValidator(Joi),
  'date-of-birth': validation.contact.createBirthDateValidator(Joi),
  postcode: validation.contact.createUKPostcodeValidator(Joi)
}).options({ abortEarly: false, allowUnknown: true })

const validator = async payload => {
  const dateOfBirth = `${payload['date-of-birth-year']}-${payload['date-of-birth-month']}-${payload['date-of-birth-day']}`
  Joi.assert(
    {
      'date-of-birth': dateOfBirth,
      postcode: payload.postcode,
      referenceNumber: payload.referenceNumber
    },
    schema
  )
}

export default pageRoute(IDENTIFY.page, IDENTIFY.uri, validator, AUTHENTICATE.uri, getData)
