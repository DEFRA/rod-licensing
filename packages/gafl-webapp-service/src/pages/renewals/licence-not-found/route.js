import { LICENCE_NOT_FOUND, NEW_TRANSACTION } from '../../../uri.js'
import pageRoute from '../../../routes/page-route.js'
import { nextPage } from '../../../routes/next-page.js'
import { addLanguageCodeToUri } from '../../../processors/uri-helper.js'

export const getData = async request => {
  const mssgs = request.i18n.getCatalog()
  return {
    title: mssgs.licence_not_found_title,
    uri: {
      new: addLanguageCodeToUri(request, NEW_TRANSACTION.uri)
    }
  }
}

export default pageRoute(LICENCE_NOT_FOUND.page, LICENCE_NOT_FOUND.uri, null, nextPage, getData)
