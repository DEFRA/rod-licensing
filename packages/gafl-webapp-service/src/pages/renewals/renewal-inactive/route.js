import { RENEWAL_INACTIVE, NEW_TRANSACTION } from '../../../uri.js'
import pageRoute from '../../../routes/page-route.js'
import moment from 'moment'
import { dateDisplayFormat } from '../../../processors/date-and-time-display.js'
import { nextPage } from '../../../routes/next-page.js'
import { RENEWAL_ERROR_REASON } from '../../../constants.js'

export const getBodyMessage = (mssgs, reason, referenceNumber, validTo) => {
  switch (reason) {
    case RENEWAL_ERROR_REASON.NOT_DUE:
      return `${mssgs.renewal_inactive_not_due_1}${referenceNumber}${mssgs.renewal_inactive_not_due_2}${validTo}`
    case RENEWAL_ERROR_REASON.EXPIRED:
      return `${mssgs.renewal_inactive_not_due_1}${referenceNumber}${mssgs.renewal_inactive_has_expired_1}${validTo}${mssgs.renewal_inactive_has_expired_2}`
    case RENEWAL_ERROR_REASON.NOT_ANNUAL:
      return `${mssgs.renewal_inactive_not_due_1}${referenceNumber}${mssgs.renewal_inactive_not_annual_1}`
    default:
      return ''
  }
}

export const getData = async request => {
  const { referenceNumber, authentication } = await request.cache().helpers.status.getCurrentPermission()
  const mssgs = request.i18n.getCatalog()
  const validTo = moment(authentication.endDate).format(dateDisplayFormat)
  const reason = authentication.reason
  const bodyMessage = getBodyMessage(mssgs, reason, referenceNumber, validTo)
  return {
    bodyMessage,
    reason: authentication.reason,
    reasonCodes: RENEWAL_ERROR_REASON,
    uri: {
      new: NEW_TRANSACTION.uri
    }
  }
}

export default pageRoute(RENEWAL_INACTIVE.page, RENEWAL_INACTIVE.uri, null, nextPage, getData)
