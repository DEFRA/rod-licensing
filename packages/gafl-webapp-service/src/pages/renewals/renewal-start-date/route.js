import { ADVANCED_PURCHASE_MAX_DAYS, SERVICE_LOCAL_TIME } from '@defra-fish/business-rules-lib'
import { dateFormats, PAGE_STATE } from '../../../constants.js'
import { RENEWAL_START_DATE, LICENCE_SUMMARY } from '../../../uri.js'
import pageRoute from '../../../routes/page-route.js'
import Joi from 'joi'
import JoiDate from '@hapi/joi-date'
import moment from 'moment-timezone'
import { displayExpiryDate, cacheDateFormat } from '../../../processors/date-and-time-display.js'
import { addLanguageCodeToUri } from '../../../processors/uri-helper.js'
import { errorShimm } from '../../../handlers/page-handler.js'
import { licenceToStart } from '../../licence-details/licence-to-start/update-transaction.js'
import { ageConcessionHelper } from '../../../processors/concession-helper.js'

const JoiX = Joi.extend(JoiDate)

const validator = (payload, options) => {
  const { permission } = options.context.app.request
  const endDateMoment = moment.utc(permission.renewedEndDate).tz(SERVICE_LOCAL_TIME)
  const licenceStartDate = `${payload['licence-start-date-year']}-${payload['licence-start-date-month']}-${payload['licence-start-date-day']}`

  return Joi.assert(
    { 'licence-start-date': licenceStartDate },
    Joi.object({
      'licence-start-date': JoiX.date()
        .format(dateFormats)
        .min(endDateMoment.clone().startOf('day'))
        .max(endDateMoment.clone().add(ADVANCED_PURCHASE_MAX_DAYS, 'days'))
        .required()
    }).options({ abortEarly: false, allowUnknown: true })
  )
}

export const getLicenceToStartAndLicenceStartTime = async (result, request) => {
  const permission = await request.cache().helpers.transaction.getCurrentPermission()
  const { payload } = await request.cache().helpers.page.getCurrentPermission(RENEWAL_START_DATE.page)
  const endDateMoment = moment.utc(permission.renewedEndDate).tz(SERVICE_LOCAL_TIME)

  if (result.error) {
    await request.cache().helpers.page.setCurrentPermission(RENEWAL_START_DATE.page, { payload, error: errorShimm(result.error) })
    await request
      .cache()
      .helpers.status.setCurrentPermission({ [RENEWAL_START_DATE.page]: PAGE_STATE.error, currentPage: RENEWAL_START_DATE.page })
    return addLanguageCodeToUri(request, RENEWAL_START_DATE.uri)
  } else {
    permission.licenceStartDate = moment({
      year: payload['licence-start-date-year'],
      month: Number.parseInt(payload['licence-start-date-month']) - 1,
      day: payload['licence-start-date-day']
    }).format(cacheDateFormat)

    if (moment(permission.licenceStartDate, cacheDateFormat).isSame(endDateMoment, 'day')) {
      // If today is the day of the renewal
      if (endDateMoment.isAfter(moment().tz(SERVICE_LOCAL_TIME))) {
        // Not yet expired - set the start time accordingly
        permission.licenceStartTime = endDateMoment.hours()
        permission.licenceToStart = licenceToStart.ANOTHER_DATE
      } else {
        // Already expired - set to 30 minutes after payment.
        permission.licenceToStart = licenceToStart.AFTER_PAYMENT
        permission.licenceStartTime = null
      }
    } else {
      // Renewing in advance - set the start time to midnight
      permission.licenceToStart = licenceToStart.ANOTHER_DATE
      permission.licenceStartTime = 0
    }
  }

  ageConcessionHelper(permission)
  await request.cache().helpers.transaction.setCurrentPermission(permission)
  return addLanguageCodeToUri(request, LICENCE_SUMMARY.uri)
}

export const getData = async request => {
  const permission = await request.cache().helpers.transaction.getCurrentPermission()

  const expiryTimeString = displayExpiryDate(request, permission)
  const endDateMoment = moment.utc(permission.renewedEndDate, null, request.locale).tz(SERVICE_LOCAL_TIME)

  return {
    expiryTimeString,
    hasExpired: permission.renewedHasExpired,
    minStartDate: endDateMoment.format('DD MM YYYY'),
    maxStartDate: endDateMoment.clone().add(ADVANCED_PURCHASE_MAX_DAYS, 'days').format('DD MM YYYY'),
    advancedPurchaseMaxDays: ADVANCED_PURCHASE_MAX_DAYS
  }
}

export default pageRoute(
  RENEWAL_START_DATE.page,
  RENEWAL_START_DATE.uri,
  validator,
  request => getLicenceToStartAndLicenceStartTime(validator, request),
  getData
)
