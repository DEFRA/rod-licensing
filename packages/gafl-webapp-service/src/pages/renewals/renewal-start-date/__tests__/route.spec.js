import { getData, getLicenceToStartAndLicenceStartTime, validator } from '../route.js'
import { RENEWAL_START_DATE, LICENCE_SUMMARY } from '../../../../uri.js'
import { addLanguageCodeToUri } from '../../../../processors/uri-helper.js'
import { ageConcessionHelper } from '../../../../processors/concession-helper.js'
import { displayExpiryDate } from '../../../../processors/date-and-time-display.js'

const mockDisplayExpiryDate = Symbol('displayExpiryDate')
jest.mock('../../../../processors/date-and-time-display.js', () => ({
  cacheDateFormat: 'YYYY-MM-DD',
  displayExpiryDate: jest.fn(() => mockDisplayExpiryDate)
}))

jest.mock('../../../../processors/concession-helper.js', () => ({
  ageConcessionHelper: jest.fn()
}))

jest.mock('../../../../processors/uri-helper.js', () => ({
  addLanguageCodeToUri: jest.fn((_request, uri) => uri)
}))

jest.mock('moment-timezone', () => {
  const realMoment = jest.requireActual('moment-timezone')
  return {
    default: targetDate => {
      if (targetDate) {
        return realMoment(targetDate)
      }
      return realMoment('2023-02-10T12:00:00.000Z')
    },
    utc: realMoment.utc
  }
})

jest.mock('../../../../handlers/page-handler.js', () => ({
  errorShimm: () => {}
}))

jest.mock('../../../../routes/page-route.js')

const mockTransactionSet = jest.fn()

const getMockRequest = (permission = getSamplePermission()) => ({
  cache: () => ({
    helpers: {
      status: {
        getCurrentPermission: jest.fn(),
        setCurrentPermission: jest.fn()
      },
      page: {
        getCurrentPermission: async () => ({
          payload: {
            'licence-start-date-year': 1999,
            'licence-start-date-month': 1,
            'licence-start-date-day': 1
          }
        }),
        setCurrentPermission: jest.fn()
      },
      transaction: {
        getCurrentPermission: () => permission,
        setCurrentPermission: mockTransactionSet
      }
    }
  }),
  i18n: {
    getCatalog: () => [],
    getLocales: () => []
  },
  url: {
    search: '?lang=cy'
  }
})

const getSamplePermission = ({ renewedEndDate = '2023-01-01T00:00:00.000Z' } = {}) => ({
  licensee: {
    noLicenceRequired: {}
  },
  renewedEndDate: renewedEndDate
})

describe('getData', () => {
  beforeEach(jest.clearAllMocks)
  it('displayExpiryDate return value', async () => {
    const expiryDate = Symbol('expiryDate')
    displayExpiryDate.mockReturnValue(expiryDate)
    const expected = await getData(getMockRequest())
    expect(expected.expiryTimeString).toBe(expiryDate)
  })

  it('displayExpiryDate is called with expected arguments', async () => {
    const permission = getSamplePermission()
    const request = getMockRequest()
    await getData(request)
    expect(displayExpiryDate).toHaveBeenCalledWith(request, permission)
  })

  it('getData returns expected outputs', async () => {
    const expected = await getData(getMockRequest())
    expect(expected).toMatchSnapshot()
  })
})

describe('getLicenceToStartAndLicenceStartTime', () => {
  beforeEach(jest.clearAllMocks)

  it.only('ageConcessionHelper is called with expected arguments', async () => {
    const result = 'result'
    await getLicenceToStartAndLicenceStartTime(result, getMockRequest())
    expect(ageConcessionHelper).toHaveBeenCalled()
  })

  it.each([
    [RENEWAL_START_DATE.uri, 'fails', true],
    [LICENCE_SUMMARY.uri, 'passes', false]
  ])('addLanguageCodeToUri is called with request and %s when validation %s', async (uri, desc, error) => {
    const result = { error: error }
    const request = getMockRequest()
    await getLicenceToStartAndLicenceStartTime(result, request)
    expect(addLanguageCodeToUri).toHaveBeenCalledWith(request, uri)
  })
})

describe('licenceStartTime and licenceToStart values', () => {
  describe.each([
    ['licenceStartTime = endDateMoment.hours and licenceToStart = ANOTHER_DATE', Symbol('end hours'), 'another-date', '2023-02-10'],
    ['licenceStartTime = null and licenceToStart = AFTER_PAYMENT', null, 'after-payment', '2023-02-10'],
    ['licenceStartTime = 0 and licenceToStart = ANOTHER_DATE', 0, 'another-date', '2023-02-10']
  ])(
    'returns values of: %s', (desc, endHours, licenceToStartResult, licenceStartDate) => {
      beforeEach(jest.clearAllMocks)

      it('licenceStartTime', async () => {
        const permission = { licenceStartDate: licenceStartDate }
        const payload = { 'licence-start-date-year': 2023, 'licence-start-date-month': 2, 'licence-start-date-day': 10 }
        await getLicenceToStartAndLicenceStartTime({ error: false }, getMockRequest(payload, permission))
        expect(mockTransactionSet).toHaveBeenCalledWith(
          expect.objectContaining({
            licenceStartTime: endHours
          })
        )
      })

      it('licenceToStart', async () => {
        await getLicenceToStartAndLicenceStartTime({ error: false }, getMockRequest())
        expect(mockTransactionSet).toHaveBeenCalledWith(
          expect.objectContaining({
            licenceToStart: licenceToStartResult
          })
        )
      })
    })
})

describe('validator', () => {
  const getMockOptions = renewedEndDate => ({
    context: {
      app: {
        request: {
          permission: {
            renewedEndDate
          }
        }
      }
    }
  })

  // it('validation fails', () => {
  //   return expect(validator({ 'licence-start-date-year': 1990, 'licence-start-date-month': 2, 'licence-start-date-day': 1 },
  //     getMockOptions('1990-02-01'))).toThrow()
  // })

  it('validation succeeds', () => {
    return expect(validator({ 'licence-start-date-year': 1990, 'licence-start-date-month': 2, 'licence-start-date-day': 1 },
      getMockOptions('1990-02-01'))).toBeUndefined()
  })
})
