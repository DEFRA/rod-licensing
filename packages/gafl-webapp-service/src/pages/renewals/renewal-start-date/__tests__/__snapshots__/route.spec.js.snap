// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getData getData returns expected outputs 1`] = `
Object {
  "advancedPurchaseMaxDays": 30,
  "expiryTimeString": Symbol(expiryDate),
  "hasExpired": undefined,
  "maxStartDate": "31 01 2023",
  "minStartDate": "01 01 2023",
}
`;
